File: Cloud Storage System/Level 3/Python/file_service.py
# file_service.py

from file import File
from file_repository import FileRepository
from user_repository import UserRepository

class FileService:
    def __init__(self, file_repo: FileRepository, user_repo: UserRepository):
        self.file_repo = file_repo
        self.user_repo = user_repo

    def add_user(self, user_id: str, capacity: int) -> str:
        return "true" if self.user_repo.add_user(user_id, capacity) else "false"

    def add_file(self, user_id: str, path: str, size: int) -> str:
        user = self.user_repo.get_user(user_id)
        if not user:
            return ""

        if self.file_repo.exists(path):
            old_file = self.file_repo.get(path)
            if self.file_repo.get_owner(path) != user_id:
                return ""
            user.delete_file(path, old_file.size)

        if not user.has_capacity(size):
            return ""

        new_file = File(path, size)
        self.file_repo.add(new_file, user_id)
        user.add_file(path, size)
        return str(user.get_remaining_capacity())

    def delete_file(self, path: str) -> str:
        entry = self.file_repo.delete(path)
        if not entry:
            return ""
        file, owner_id = entry
        owner = self.user_repo.get_user(owner_id)
        if owner:
            owner.delete_file(path, file.size)
        return str(file.size)

    def get_file_size(self, path: str) -> str:
        file = self.file_repo.get(path)
        return str(file.size) if file else ""

    def get_largest(self, prefix: str, n: int) -> str:
        all_files = [file for (file, _) in self.file_repo.get_all() if file.path.startswith(prefix)]
        if not all_files:
            return ""
        sorted_files = sorted(all_files, key=lambda f: (-f.size, f.path))
        top_files = sorted_files[:n]
        return ", ".join(f"{f.path}({f.size})" for f in top_files)

    def merge_user(self, user1_id: str, user2_id: str) -> str:
        result = self.user_repo.merge_users(user1_id, user2_id)
        return str(result) if result is not None else ""

File: Cloud Storage System/Level 3/C#/file_service.cs
// FileService.cs
using System.Linq;

public class FileService
{
    private readonly FileRepository _fileRepo;
    private readonly UserRepository _userRepo;

    public FileService(FileRepository fileRepo, UserRepository userRepo)
    {
        _fileRepo = fileRepo;
        _userRepo = userRepo;
    }

    public string AddUser(string userId, int capacity)
    {
        return _userRepo.AddUser(userId, capacity) ? "true" : "false";
    }

    public string AddFile(string userId, string path, int size)
    {
        var user = _userRepo.GetUser(userId);
        if (user == null) return "";

        if (_fileRepo.Exists(path))
        {
            var oldFile = _fileRepo.Get(path);
            var owner = _fileRepo.GetOwner(path);
            if (owner != userId) return "";
            user.DeleteFile(path, oldFile!.Size);
        }

        if (!user.HasCapacity(size)) return "";

        var newFile = new File(path, size);
        _fileRepo.Add(newFile, userId);
        user.AddFile(path, size);

        return user.RemainingCapacity().ToString();
    }

    public string DeleteFile(string path)
    {
        var entry = _fileRepo.Delete(path);
        if (entry == null) return "";

        var (file, ownerId) = entry.Value;
        var owner = _userRepo.GetUser(ownerId);
        owner?.DeleteFile(path, file.Size);
        return file.Size.ToString();
    }

    public string GetFileSize(string path)
    {
        var file = _fileRepo.Get(path);
        return file != null ? file.Size.ToString() : string.Empty;
    }

    public string GetLargest(string prefix, int n)
    {
        var allFiles = _fileRepo.GetAll()
            .Where(x => x.Item1.Path.StartsWith(prefix))
            .OrderByDescending(x => x.Item1.Size)
            .ThenBy(x => x.Item1.Path)
            .Take(n)
            .Select(x => $"{x.Item1.Path}({x.Item1.Size})")
            .ToList();

        return allFiles.Count == 0 ? string.Empty : string.Join(", ", allFiles);
    }

    public string MergeUser(string user1Id, string user2Id)
    {
        var result = _userRepo.MergeUsers(user1Id, user2Id);
        return result.HasValue ? result.Value.ToString() : string.Empty;
    }
}
